chain/chain/src/store/mod.rs

```diff
@@ -712,6 +720,54 @@ impl ChainStore {
             }
         }
     }
+
+    pub fn compute_transaction_validity(
+        &self,
+        protocol_version: ProtocolVersion,
+        prev_block_header: &BlockHeader,
+        chunk: &ShardChunk,
+    ) -> Result<Vec<bool>, Error> {
+        let relaxed_chunk_validation = near_primitives::checked_feature!(
+            "protocol_feature_relaxed_chunk_validation",
+            RelaxedChunkValidation,
+            protocol_version
+        );
+        if near_primitives::checked_feature!("stable", AccessKeyNonceRange, protocol_version) {
+            let tx_iter = chunk.transactions().into_iter();
+            if relaxed_chunk_validation {
+                let valid_txs = tx_iter
+                    .map(|transaction| {
+                        let tx_valid = self
+                            .check_transaction_validity_period(
+                                prev_block_header,
+                                transaction.transaction.block_hash(),
+                            )
+                            .is_ok();
+                        tx_valid
+                    })
+                    .collect::<Vec<bool>>();
+                Ok(valid_txs)
+            } else {
+                tx_iter
+                    .map(|transaction| {
+                        let tx_valid = self
+                            .check_transaction_validity_period(
+                                prev_block_header,
+                                transaction.transaction.block_hash(),
+                            )
+                            .is_ok();
+                        if !tx_valid {
+                            Err(Error::from(Error::InvalidTransactions))
+                        } else {
+                            Ok(true)
+                        }
+                    })
+                    .collect::<Result<Vec<bool>, _>>()
```

Comment:
nit: that is the part that I mentioned to be quite a challenge to understand: collecting into `Result<..>`. Previously this was implemented as a simple `for` loop with early return on error. Feel free to keep this, just wanted to share my perspective.

Reply:
Fair enough, will restore it to a loop. `collect::<Result<_, _>>` is a fairly common idiom (and we have some of them across the code-base.) Might be worthwhile to add a style guide entry on this or something :thinking:.